src/goto-programs/remove_virtual_functions.cpp

```diff
@@ -144,6 +146,63 @@ static void create_static_function_call(
   }
 }
 
+/// If \p argument_for_this contains a dereference then create a temporary
+/// variable for it and use that instead
+/// \param [in,out] argument_for_this: The first argument of the function call
+/// \param symbol_table: The symbol table to add the new symbol to
+/// \param vcall_source_loc: The source location of the function call, which is
+///   used for new instructions that are added
+/// \param [out] new_code_for_this_argument: New instructions are added here
+static void process_this_argument(
+  exprt &argument_for_this,
+  symbol_table_baset &symbol_table,
+  const source_locationt &vcall_source_loc,
+  goto_programt &new_code_for_this_argument)
+{
+  const auto dereference_it = std::find_if(
+    argument_for_this.depth_cbegin(),
+    argument_for_this.depth_cend(),
+    [](const exprt &expr) { return can_cast_expr<dereference_exprt>(expr); });
+
+  const bool argument_for_this_contains_a_dereference =
+    dereference_it != argument_for_this.depth_cend();
+
+  if(argument_for_this_contains_a_dereference)
+  {
+    // Create a temporary for the `this` argument. This is so that value-set
+    // filtering in symex will reduce the value-set for `this` to those with the
+    // correct class identifier.
+    symbolt &temp_symbol = get_fresh_aux_symbol(
+      argument_for_this.type(),
+      "remove_virtual_functions", // TODO: pass function name in to use here?
+      "this_argument",
+      vcall_source_loc,
+      ID_java, //TODO: get this from somewhere
+      symbol_table);
+    const symbol_exprt temp_var_for_this = temp_symbol.symbol_expr();
+
+    new_code_for_this_argument.add(
+      goto_programt::make_decl(temp_var_for_this, vcall_source_loc));
+    new_code_for_this_argument.add(
+      goto_programt::make_assignment(
+        temp_var_for_this, argument_for_this, vcall_source_loc));
+
+    // Add null check for temp_var_for_this, as the old one won't work any more. We should
+    // really use java_bytecode_instrumentt::check_null_dereference(), but it is
+    // in jbmc rather than cbmc. Or we could use create_fatal_assertion(), but it
+    // returns a code_blockt, which is a bit inconvenient.
+    const notequal_exprt null_check(
+      temp_var_for_this,
+      null_pointer_exprt(to_pointer_type(temp_var_for_this.type())));
+    new_code_for_this_argument.add(
+      goto_programt::make_assertion(null_check, vcall_source_loc));
+    new_code_for_this_argument.add(
+      goto_programt::make_assumption(null_check, vcall_source_loc));
```

Comment:
That's only valid for Java, is it?

Reply:
Hmm. In java we know there must have been a null check before the virtual function call. I put this in because without it the local safe pointer analysis wasn't realising that `this` definitely wasn't null after the null check. Other options include:
1) making the local safe pointer analysis a teensy bit smarter;
2) checking for a null check in the two instructions above the virtual function call, and only inserting a null check here if we find one.

Reply:
I've done option 2.